packages/integration-sdk-cli/src/neo4j/neo4jGraphStore.ts

```diff
@@ -0,0 +1,135 @@
+      return result;
+    }
+  }
+
+  private startsWithNumeric(str: string){
+    return /^\d+$/.test(str);
+  }
+
+  private sanitizePropertyName(propertyName: string): string {
+    let sanitizedName = '';
+    if(this.startsWithNumeric(propertyName)) {
+      sanitizedName += 'n';
+    }
+    sanitizedName += propertyName;
+    sanitizedName = sanitizedName.replace(/[\s!@#$%^&*()-=+\\|'";:/?.,><`~\t\n[\]{}]/g, "_");
+    return sanitizedName;
+  }
+
+  private buildPropertyString(propList, nodeName) {
+    let propString = '';
+    for (const key in propList) {
+      if (key === '_rawData') {
+        //stringify JSON in rawData so we can store it.
+        propString += `${nodeName}.${key} = '${JSON.stringify(propList[key])}', `;
+      } else {
+        // Escape single quotes so they don't terminate strings prematurely
+        const finalValue = propList[key].toString().replace(/'/gi, "\\'");
+        // Sanitize out characters that aren't allowed in property names
+        const propertyName = this.sanitizePropertyName(key);
+        propString += `${nodeName}.${propertyName} = '${finalValue}', `;
+      }
+    }
+    propString = propString.slice(0, -2);
+
+    return propString;
+  }
+
+  async addEntities(newEntities: Entity[]) {
+    for await (const entity of newEntities) {
+      //Add constraint if not already in types
+      //We check for existence in case we're ever running in an instance with
+      //multiple inputs.
+      if (!this.typeList.has(entity._type)) {
+        await this.runCypherCommand(
+          `CREATE CONSTRAINT unique_${entity._type} IF NOT EXISTS ON (n:${entity._type}) ASSERT n._key IS UNIQUE;`,
+        );
+        this.typeList.add(entity._type);
+      }
+      const propString = this.buildPropertyString(entity, 'n');
+
+      const buildCommand = `MERGE (n:${entity._type} {_key: '${entity._key}'}) SET ${propString};`;
+      await this.runCypherCommand(buildCommand);
+    }
+    return Promise.resolve();
+  }
+
+  async addRelationships(newRelationships: Relationship[]) {
+    for await (const relationship of newRelationships) {
+      const propString = this.buildPropertyString(relationship, 'relationship');
+
+      //Get start and end _keys.  Will be overwritten if we're
+      //working with a mapped relationship.
+      let startEntityKey = relationship._fromEntityKey;
+      let endEntityKey = relationship._toEntityKey;
+
+      if(relationship._mapping) { //Mapped Relationship
+        if(relationship._mapping['skipTargetCreation'] === false) {
+          //Create target entity first
+          const tempEntity: Entity = {
+            _class: relationship._mapping['targetEntity']._class,
+            //TODO, I think this key is wrong, but not sure what else to use
+            _key:  relationship._key.replace(relationship._mapping['sourceEntityKey'], ''),
+            _type: relationship._mapping['targetEntity']._type,
+          }
+          await this.addEntities([tempEntity]);
```

Comment:
The addEntities command is performing a MERGE on the neo4j database, which will handle any collisions.

Reply:
And the MERGE effectively does a MATCH followed by a CREATE, so we're not really wasting any database clock cycles in letting it handle things, since we'd have to perform a MATCH ourselves when checking for existence.

Reply:
Thanks for the explanation, @adam-in-ict ! TIL!